plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs

```diff
@@ -0,0 +1,135 @@
+    (ns, st)
+  where
+    nameSemanticFromRefMap :: RefMap a -> Name -> Maybe HsSemanticTokenType
+    nameSemanticFromRefMap rm' name' = do
+      spanInfos <- -- traceShow ("getting spans:", nameString) $
+        Map.lookup (Right name') rm'
+      let infos = S.unions $ map (identInfo . snd) spanInfos
+      let typeTokenType = fmap (typeSemantic hieKind) $ listToMaybe $ mapMaybe (identType . snd) spanInfos
+      let contextInfoTokenType = contextInfosMaybeTokenType infos
+      maximum <$> NE.nonEmpty (catMaybes [typeTokenType, contextInfoTokenType])
+
+    contextInfosMaybeTokenType :: Set.Set ContextInfo -> Maybe HsSemanticTokenType
+    contextInfosMaybeTokenType details = case NE.nonEmpty $ Set.toList details of
+      Just infos -> maximum $ NE.map infoTokenType infos
+      Nothing    -> Nothing
+
+-----------------------------------
+
+-- * extract location from HieAST a
+
+-----------------------------------
+
+hieAstSpanNames :: NameSet -> HieAST a -> [(Range, Name)]
+hieAstSpanNames nameSet ast =
+  if null (nodeChildren ast)
+    then getIds ast
+    else concatMap (hieAstSpanNames nameSet) (nodeChildren ast)
+  where
+    getIds :: HieAST a -> [(Range, Name)]
+    getIds ast' =
+      [ (realSrcSpanToRange $ nodeSpan ast', c)
+        | (Right c) <- S.toList $ getNodeIds' ast,
+          elemNameSet c nameSet
+      ]
+    getNodeIds' :: HieAST a -> S.Set Identifier
+    getNodeIds' =
+      Map.foldl' combineNodeIds mempty
+        . Map.filterWithKey (\k _ -> k == SourceInfo)
+        . getSourcedNodeInfo
+        . sourcedNodeInfo
+    combineNodeIds :: S.Set Identifier -> NodeInfo a -> S.Set Identifier
+    ad `combineNodeIds` (NodeInfo _ _ bd) = ad `S.union` xs
+      where
+        xs = M.keysSet bd
+
+-------------------------------------------------
+
+-- * extract semantic tokens from NameSemanticMap
+
+-------------------------------------------------
+
+semanticTokenAbsoluteSemanticTokens :: [(Range, HsSemanticTokenType)] -> Either Text SemanticTokens
+semanticTokenAbsoluteSemanticTokens = makeSemanticTokens defaultSemanticTokensLegend . List.sort . map (uncurry toAbsSemanticToken)
+  where
+    toAbsSemanticToken :: Range -> HsSemanticTokenType -> SemanticTokenAbsolute
+    toAbsSemanticToken (Range (Position startLine startColumn) (Position _endLine endColumn)) tokenType =
+      let len = endColumn - startColumn
+       in SemanticTokenAbsolute
+            (fromIntegral startLine)
+            (fromIntegral startColumn)
+            (fromIntegral len)
+            (toLspTokenType tokenType)
+            []
+
+extractSemanticTokensFromNames :: NameSemanticMap -> [(Range, Name)] -> [(Range, HsSemanticTokenType)]
+extractSemanticTokensFromNames nsm = mergeNameFromSamSpan . mapMaybe (getSemantic nsm)
+  where
+    -- merge all tokens with same span
+    -- mergeNameFromSamSpan :: [(Span, SemanticTokenType)] -> [(Span, SemanticTokenType)]
+    mergeNameFromSamSpan xs = Map.toList $ Map.fromListWith (<>) xs
```

Comment:
Might be cleaner to just work with a `Map Range X` as much as possible until the last minute when we need a list? X could be `Set Name` or something until this point.

Reply:
good idea